server/src/main/java/org/opensearch/common/cache/stats/ImmutableCacheStatsHolder.java

```diff
@@ -69,15 +153,173 @@ public ImmutableCacheStats getStatsForDimensionValues(List<String> dimensionValu
         return current.stats;
     }
 
+    /**
+     * Returns a new tree containing the stats aggregated by the levels passed in.
+     * The new tree only has dimensions matching the levels passed in.
+     * The levels passed in must be in the proper order, as they would be in the output of filterLevels().
+     */
+    Node aggregateByLevels(List<String> filteredLevels) {
+        if (filteredLevels.isEmpty()) {
+            throw new IllegalArgumentException("Filtered levels passed to aggregateByLevels() cannot be empty");
+        }
+        Node newRoot = new Node("", false, statsRoot.getStats());
+        for (Node child : statsRoot.children.values()) {
+            aggregateByLevelsHelper(newRoot, child, filteredLevels, 0);
+        }
+        return newRoot;
+    }
+
+    /**
+     * Because we may have to combine nodes that have the same dimension name, I don't think there's a clean way to aggregate
+     * fully recursively while also passing in a completed map of children nodes before constructing the parent node.
+     * For this reason, in this function we have to build the new tree top down rather than bottom up.
+     * We use private methods allowing us to add children to/increment the stats for an existing node.
+     * This should be ok because the resulting trees are short-lived objects that are not exposed anywhere outside this class,
+     * and the original tree is never changed.
+     */
+    private void aggregateByLevelsHelper(Node parentInNewTree, Node currentInOriginalTree, List<String> levels, int depth) {
+        if (levels.contains(dimensionNames.get(depth))) {
+            // If this node is in a level we want to aggregate, create a new dimension node with the same value and stats, and connect it to
+            // the last parent node in the new tree. If it already exists, increment it instead.
+            String dimensionValue = currentInOriginalTree.getDimensionValue();
+            Node nodeInNewTree = parentInNewTree.children.get(dimensionValue);
+            if (nodeInNewTree == null) {
+                // Create new node with stats matching the node from the original tree
+                int indexOfLastLevel = dimensionNames.indexOf(levels.get(levels.size() - 1));
+                boolean isLeafNode = depth == indexOfLastLevel; // If this is the last level we aggregate, the new node should be a leaf
+                // node
+                nodeInNewTree = new Node(dimensionValue, isLeafNode, currentInOriginalTree.getStats());
+                parentInNewTree.addChild(dimensionValue, nodeInNewTree);
```

Comment:
Doing all this logic as part of ToXContent method will be overhead for the coordinator node while returning response. We should flow the level parameter as part of the NodeStatsRequest and only construct, serialize what we actually need